maestro-cli/src/main/java/maestro/cli/cloud/CloudInteractor.kt

```diff
@@ -0,0 +1,203 @@
+                uploadProgress.set(totalBytes, bytesWritten)
+            }
+            println()
+
+            if (async) {
+                PrintUtils.message("✅ Upload successful! View the results of your upload below:")
+                PrintUtils.message(uploadUrl(uploadId, teamId, appId))
+
+                return 0
+            } else {
+                PrintUtils.message("Waiting for analyses to complete...")
+                println()
+
+                return waitForCompletion(
+                    authToken = authToken,
+                    uploadId = uploadId,
+                    teamId = teamId,
+                    appId = appId,
+                )
+            }
+        }
+    }
+
+    private fun waitForCompletion(
+        authToken: String,
+        uploadId: String,
+        teamId: String,
+        appId: String,
+    ): Int {
+        val startTime = System.currentTimeMillis()
+
+        val reportedCompletions = mutableSetOf<String>()
+
+        var pollingInterval = minPollIntervalMs
+        var retryCounter = 0
+        do {
+            val upload = try {
+                client.uploadStatus(authToken, uploadId)
+            } catch (e: ApiClient.ApiException) {
+                if (e.statusCode == 429) {
+                    // back off through extending sleep duration with 25%
+                    pollingInterval = (pollingInterval * 1.25).toLong()
+                    Thread.sleep(pollingInterval)
+                    continue
+                }
+
+                if (e.statusCode == 500) {
+                    if (++retryCounter <= maxPollingRetries) {
+                        // retry on 500
+                        Thread.sleep(pollingInterval)
+                        continue
+                    }
+                }
+
+                throw CliError("Failed to fetch the status of an upload $uploadId. Status code = ${e.statusCode}")
+            }
+
+            upload.flows
+                .filter {
+                    it.name !in reportedCompletions && it.status in COMPLETED_STATUSES
+                }
+                .forEach {
+                    UploadStatusView.showFlowCompletion(it)
+
+                    reportedCompletions.add(it.name)
+                }
+
+            if (upload.completed) {
+                UploadStatusView.showUploadResult(upload, teamId, appId)
+
+                return if (upload.status == UploadStatus.Status.ERROR) {
+                    1
+                } else {
+                    0
+                }
+            }
+
+            Thread.sleep(pollingInterval)
+        } while (System.currentTimeMillis() - startTime < waitTimeoutMs)
+
+        PrintUtils.warn("Upload did not complete in time due to an issue on mobile.dev side. Follow the results of your upload below:")
+        println(uploadUrl(uploadId, teamId, appId))
+
+        return if (failOnTimeout) {
+            PrintUtils.err("FAIL")
+
+            1
+        } else {
+            PrintUtils.warn("SKIPPED")
```

Comment:
What does skipped mean exactly?

Reply:
This is a rare case where our backend did not complete the upload in 30 minutes. In such case, we are not going to fail the CLI command and instead report it as skipped (exit code 0)

Reply:
Got it. At first glance, the logic feels like "skipped" is a catch all and I just want to make sure that we aren't letting real issues fall through the cracks by bucketing errors into this skipped state

Reply:
We are going to retry  a few times and then eventually throw an error in case of server errors: https://github.com/mobile-dev-inc/maestro/pull/345/files#diff-c0b9059374eeada60a2aa7a5d016bcbd26ce2fee9e77ce674648c4d9f08550c3R134

Reply:
Ok reading the code more carefully now this makes perfect sense. Thanks!